File: find_solutions.py
from typing import List
from tqdm import tqdm
from time import perf_counter
import csv

ALPHABET = ["q", "x", "j", "z", "v", "f", "w", "b", "k", "g", "p", "m", "h",
            "d", "c", "y", "t", "l", "n", "u", "r", "o", "i", "s", "e", "a"]
INPUTFILE = "words_alpha.txt"
OUTPUTFILE = "solutions.csv"


def word_has_unique_letters(word: str):
    """Checks if a word contains unique letters"""
    for i, letter_1 in enumerate(word[:-1]):
        for letter_2 in word[i+1:]:
            if letter_1 == letter_2:
                return False
    return True


def words_have_unique_letters(words: List[str], new_word: str):
    """Checks if the letters in two words of the same length are unique."""
    for old_word in words:
        for letter_1 in old_word:
            for letter_2 in new_word:
                if letter_1 == letter_2:
                    return False
    return True


def first_letter_in_word(word: str):
    for letter in ALPHABET:
        if letter in word:
            break
    return letter


def remove_letters_in_solution(solution):
    letters = ALPHABET.copy()
    for word in solution:
        for letter in word:
            letters.remove(letter)
    return letters


def read_input_to_dict(file: str):
    """
    Reads the words from the input file.
    For each letter from ALPHABET there is a dictionary item.
    Each word is assigned to the first letter in ALHPABET it contains.
    All words which contain a q are added to "q", all words which contain an "x" but not a "q" to "x", etc..
    """
    output = {}
    for letter in ALPHABET:
        output[letter] = []

    with open(file) as f:
        for i, line in enumerate(f.readlines()):
            word = line[:-1]
            if len(word) == 5:
                if word_has_unique_letters(word):
                    output[first_letter_in_word(word)].append(word)
    return output


def find_solutions(words):
    solutions = []

    for i_1, letter_1 in enumerate(ALPHABET[:2]):
        print(f"Finding solutions with a word containing the letter {letter_1} (iteration {i_1 + 1}/2)")

        for word_1 in tqdm(words[letter_1]):
            solution = [word_1]
            remaining_letters = remove_letters_in_solution(solution)

            for i_2, letter_2 in enumerate(remaining_letters[:2]):
                if i_1 == 1 and i_2 == 0:
                    continue

                for word_2 in words[letter_2]:
                    solution = [word_1]
                    if words_have_unique_letters(solution, word_2):
                        solution = [word_1, word_2]
                        remaining_letters = remove_letters_in_solution(solution)

                    for i_3, letter_3 in enumerate(remaining_letters[:2]):
                        if len(solution) < 2:
                            break

                        if i_2 == 1 and i_3 == 0:
                            continue

                        for word_3 in words[letter_3]:
                            solution = [word_1, word_2]
                            if words_have_unique_letters(solution, word_3):
                                solution = [word_1, word_2, word_3]
                                remaining_letters = remove_letters_in_solution(solution)

                            for i_4, letter_4 in enumerate(remaining_letters[:2]):
                                if len(solution) < 3:
                                    break

                                if i_3 == 1 and i_4 == 0:
                                    continue

                                for word_4 in words[letter_4]:
                                    solution = [word_1, word_2, word_3]
                                    if words_have_unique_letters(solution, word_4):
                                        solution = [word_1, word_2, word_3, word_4]
                                        remaining_letters = remove_letters_in_solution(solution)

                                    for i_5, letter_5 in enumerate(remaining_letters[:2]):
                                        if len(solution) < 4:
                                            break

                                        if i_4 == 1 and i_5 == 0:
                                            continue

                                        for word_5 in words[letter_5]:
                                            solution = [word_1, word_2, word_3, word_4]
                                            if words_have_unique_letters(solution, word_5):
                                                solution = [word_1, word_2, word_3, word_4, word_5]
                                                solutions.append(solution)
    return solutions


def save_solutions(filename: str, solutions: List[List[str]]):
    with open(filename, "w") as f:
        writer = csv.writer(f, delimiter=',')
        for solution in solutions:
            writer.writerow(solution)


def main():
    t1 = perf_counter()
    words = read_input_to_dict(INPUTFILE)

    solutions = find_solutions(words)
    save_solutions(OUTPUTFILE, solutions)
    t2 = perf_counter()

    print(f"Found {len(solutions)} solutions, took {t2-t1:.3f} seconds.")

if __name__ == "__main__":
    main()

File: find_solutions.kt
import kotlin.time.*

val ALPHABET = "qxjzvfwbkgpmhdcytlnuroisea"
val INPUTFILE = "words_alpha.txt"
val OUTPUTFILE = "solutions.csv"

fun wordHasUniqueLetters(word: String): Boolean{
    // Checks if the letters in a word are unique
    for (i in 0..<word.length-1){
        if (word.get(i) in word.subSequence(i+1, word.length)){
            return false
        }
    }
    return true
}

fun wordsHaveUniqueLetters(words: List<String>, word: String): Boolean {
    for (oldWord in words) {
        for (letter in oldWord){
            if (letter in word) {
                return false
            }
        }
    }
    return true
}

fun firstLetterInWord(word: String): Char {
    var firstLetter = '_'
    for (letter in ALPHABET) {
        if (letter in word) {
            firstLetter = letter
            break
        }
    }
    return firstLetter
}

fun removeLettersInSolutionFromAlphabet(solution: List<String>): String {
    var remaining_letters = ALPHABET
    for (word in solution) {
        for (letter in word) {
            remaining_letters = remaining_letters.replace(letter.toString(), "")
        }
    }
    return remaining_letters
}

fun readInputToMap(file: String): Map<Char, ArrayList<String>> {
    var result: MutableMap<Char, ArrayList<String>> = mutableMapOf()

    for (letter in ALPHABET) {
        var emptyList: ArrayList<String> = arrayListOf()
        result.put(letter, emptyList)
    }

    val file = java.io.File(file)
    file.bufferedReader().forEachLine {
        val word = it
        if (word.length == 5 && wordHasUniqueLetters(word)) {
            result[firstLetterInWord(word)]?.add(word)
        }
    }
    return result
}

fun findSolutions(words: Map<Char, ArrayList<String>>): List<List<String>> {
    var solutions: MutableList<List<String>> = mutableListOf()

    for (i1 in 0..1) {
        val letter1 = ALPHABET[i1]
        println("Finding solutions with a word containing the letter $letter1 (iteration ${i1 + 1}/2)")

        for (word1 in words[letter1].orEmpty()) {
            var solution = mutableListOf(word1)
            var remaining_letters = removeLettersInSolutionFromAlphabet(solution)

            for (i2 in 0..1) {
                if (i1==1 && i2==0) { continue }
                val letter2 = remaining_letters[i2]

                for (word2 in words[letter2].orEmpty()) {
                    solution = mutableListOf(word1)

                    if (wordsHaveUniqueLetters(solution, word2)) {
                        solution.add(word2)
                        remaining_letters = removeLettersInSolutionFromAlphabet(solution)
                    }

                    for (i3 in 0..1) {
                        if (solution.size != 2) { break }
                        if (i2==1 && i3==0) { continue }
                        val letter3 = remaining_letters[i3]
                        for (word3 in words[letter3].orEmpty()) {
                            solution = mutableListOf(word1, word2)

                            if (wordsHaveUniqueLetters(solution, word3)) {
                                solution.add(word3)
                                remaining_letters = removeLettersInSolutionFromAlphabet(solution)
                            }

                            for (i4 in 0..1) {
                                if (solution.size != 3) { break }
                                if (i3==1 && i4==0) { continue }
                                val letter4 = remaining_letters[i4]

                                for (word4 in words[letter4].orEmpty()) {
                                    solution = mutableListOf(word1, word2, word3)
                                    if (wordsHaveUniqueLetters(solution, word4)) {
                                        solution.add(word4)
                                        remaining_letters = removeLettersInSolutionFromAlphabet(solution)
                                    }

                                    for (i5 in 0..1) {
                                        if (solution.size != 4) { break }
                                        if (i4==1 && i5==0) { continue }
                                        val letter5 = remaining_letters[i5]

                                        for (word5 in words[letter5].orEmpty()) {
                                            solution = mutableListOf(word1, word2, word3, word4)
                                            if (wordsHaveUniqueLetters(solution, word5)) {
                                                solution.add(word5)
                                                solutions.add(solution)
                                            }
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
    return solutions
}

fun saveSolutions(filePath: String, solutions: List<List<String>>) {
    val fileWriter = java.io.File(filePath).bufferedWriter()

    fileWriter.use { out ->
        for (solution in solutions) {
            val line = solution.joinToString(separator=",")
            out.write(line)
            out.newLine()
        }
    }
}

fun main(){
    val timeSource = TimeSource.Monotonic
    val t1 = timeSource.markNow()

    var wordsMap = readInputToMap(INPUTFILE)
    val solutions = findSolutions(wordsMap)
    saveSolutions(OUTPUTFILE, solutions)

    val t2 = timeSource.markNow()

    println("Found ${solutions.size} solutions, took ${t2-t1}.")

}
